lib/ansible/module_utils/common/hashing.py

```diff
@@ -0,0 +1,152 @@
+from __future__ import annotations
+
+import hashlib
+import os
+
+from hashlib import sha1
+
+try:
+    from hashlib import md5 as _md5
+except ImportError:
+    # Assume we're running in FIPS mode here
+    _md5 = None  # type: ignore[assignment]
+
+from ansible.module_utils.common.text.converters import to_bytes
+from ansible.module_utils.six import string_types
+
+
+def _get_available_hash_algorithms():
+    """Return a dictionary of available hash function names and their associated function."""
+    algorithms = {}
+    for algorithm_name in hashlib.algorithms_available:
+        algorithm_func = getattr(hashlib, algorithm_name, None)
+        if algorithm_func:
+            try:
+                # Make sure the algorithm is actually available for use.
+                # Not all algorithms listed as available are actually usable.
+                # For example, md5 is not available in FIPS mode.
+                algorithm_func()
+            except Exception:
+                pass
+            else:
+                algorithms[algorithm_name] = algorithm_func
+
+    return algorithms
+
+
+AVAILABLE_HASH_ALGORITHMS = _get_available_hash_algorithms()
+
+
+def generate_secure_checksum(data, hash_func='sha256'):
+    """Generates a secure checksum for the given data using the specified hash function.
+
+    Args:
+        data: The data to be hashed.
+        hash_func: The hash function to use (default: sha256).
+
+    Returns:
+        A hexadecimal string representing the checksum.
+    """
+    if hash_func is None:
+        raise ValueError("The parameter 'hash_func' can not be None")
+
+    if isinstance(hash_func, string_types):
+        if hash_func not in AVAILABLE_HASH_ALGORITHMS:
+            raise ValueError(f"{hash_func} is not available. Available algorithms: {', '.join(AVAILABLE_HASH_ALGORITHMS)}")
+        hash_func = getattr(hashlib, hash_func, None)
+
+    if not callable(hash_func):
+        raise ValueError("The parameter value of 'hash_func' is not callable. Please make sure hash_func is either string or method.")
+
+    digest = hash_func()
+    data = to_bytes(data, errors='surrogate_or_strict')
+    digest.update(data)
+    return digest.hexdigest()
+
+
+def generate_secure_file_checksum(filename, hash_func='sha256', write_to=None):
+    """Return a secure hash hex digest of local file, None if file is not present or a directory.
+
+    Args:
+        filename: The filename to be hashed.
+        hash_func: The hash function to use (default: sha256).
+        write_to: The file handle to write to (default: None).
+
+    Returns:
+        A hexadecimal string representing the checksum.
+    """
+    b_filename = to_bytes(filename, errors='surrogate_or_strict')
+
+    if not os.path.exists(b_filename) or os.path.isdir(to_bytes(filename, errors='strict')):
+        return None
+
+    if hash_func is None:
+        raise ValueError("The parameter 'hash_func' can not be None")
+
+    if isinstance(hash_func, string_types):
```

Comment:
instead of duplicating this part, make it a common function or just call the data function as an iterator